istresearch/rapidpro
# P4-3013 - Force channel delete, with warning

# For the Reviewer

When this is complete, you should approve the PR via github.

# For the Reviewee

<Reminder PR Title should be JIRA_NUMBER and Useful Description>

## Summary
Adjusted the v2 api and UI to allow optional forced channel deletion. By default deletions will throw a flow in use error. 

On the API, you must pass a `force_delete` boolean parameter. On the UI it will show a dialog with a warning, with a force remove button. 

#### Release Note
If a channel is in use, A warning will be shown, but it can be force deleted. 

#### Breaking Changes
<Description for Techops of how to handle changes, migrations, updates>None.

## Quality Assurance

You have gathered the following items:

Given all of the items above, you have updated your Application ORT at the following locations:
- **Features and Alerting**: <link to app ORT>Required.
- **P4 Alerting**: <link to p4-alerting PR>Required.

## Testing and Verification
UI Test

1. Create a channel
2. Attach it to a flow via the `Update Contact->Channel` node
3. Attempt to delete it. A warning should be shown, if you click `force remove` it will still delete it. 
![Screen Shot 2021-10-01 at 10 20 04 AM](https://user-images.githubusercontent.com/25488234/135635781-3fb6ceab-1371-4c4b-bb33-304f6f4c9a0e.png)

 API test 

1. Repeat steps 1-2 from above
2. Get your API key from the org page
3. Get the channel UUID from the channel's page
4. Run a curl command against the API:
```
curl --location --request DELETE 'http://localhost:8001/ext/api/v2/channels.json?uuid={{ CHANNEL_UUID_HERE }}&force_delete=false' --header 'Authorization: Token {{ API_TOKEN_HERE }}'
```
5. You should get back an in-use error. Toggle `force_delete=true` to force it.
